getstation/tapas
# [GUILLAUME-GUARD][REACT-1] - a first implementation of drag n drop

> Here, make a summary of what challenges, or
what parts of challenges, you were able to
complete, or not.

i was able to complete everything but the animation (i've simple added opacity effect to the list item im dragging)

> Give an estimation of time spent on each.

consuming the endpoint (1h  i had to read about how the hooks works exactly to be able to make an axios http get request) 
css ( 30 mins , i had to reead the doc about the react-jss)
Add a "loading" indicator for when the users are being fetched (3 mins, im used to this, just had to use it as a hook),
You should be able to drag and drop elements of your list. To ease this part, you must use React DND( this was the core, i've sepnt more than 3 hours reading, watching examples, and trying to understand how to adapt it with hooks)
animation ( 5 mins)


## Test instructions and expectations

> Provide the reviewer with instructions to
test your implementation and with what to
expect when testing.

clone the branch, run npm install and just just 'npm run start'
you expect to see a fetched list on the screen, and each element of this list is drag n droppable to another position in the list, pushing other elements up or down on its way.

## How does it work?

> Explain the key elements of your solutions.
Link to documentations if necessary.

so we're here implementing functional react with hooks. we first fetch the data with an axios get request, and store in the main component state the data previously fetched. thanks to this data, we generate a list of componenets that i've named Cards. When a card is dragged, it will swap positions with the card on top or down of it depending on the direction we're draging it (we calculate using the current mouse position)

### Alternatives

> If any, list alternative solutions, valid
or not, and explain your choice.

there must be a lot of alternatives to the drag n drop logic, but as i've found a working example that was pretty much solving the problem, i decided to use that one.
for the animation, it was imposible to learn everything needed to implement it in the time given, but i did find a solution that can be implemented in my case:
[https://medium.com/@andrew_b_berg/using-react-motion-with-react-dnd-85755f2ecc68](url)

this use the motion library to add the animation effect to our DND wrapped component, it was definitivly what i needed but as said previously , time was too short

> If you considered or tried other solutions
that did not work, it's a good opportunity
to mention them.

check the comment above about the animation solution.


## Critical analysis

### General feeling

> What have you thought about these challenges?
>
well, this react challenge was super entrataining, i've learn a lot about functional react, and i've discovered this Drag n drop lib.

> Do you think your implementation reflects
correctly your technical skills?

well, not really. I've always been coding OOP react , and functional react is a total new way of thinking . in those 4 and a half hours, i've been more learning that actually showing my react skills. we will clearly need to discuss this orally :p
> (we'll have the opportunity to discuss these
points orally, if necessary)

### Difficulties

> If any, list the difficulties you faced
when trying to complete the challenges and
how you overcome them.
>
> (you very probably faced difficulties)

the main difficulty is the functional react way of thinking, its something new to me, i've always been coding in OOP react. i've never used this dnd library neither. hooks may ease the use of react, but they are way harder to understand than OOP . But after this challenge, i've seen i still have a lot to learn and i definitvly want to level up in functional react :dagger: 

### Strengths and weaknesses

> List and argue the points of strengths
or weaknesses/oddnesses of your implementation.
>
well, its hard to tell what are the strenghts of the code. it does what's it meant to, i guess that already a succes.
> For weaknesses/oddnesses, suggest improvements
and give an estimation of the time
necessary to implement them.
>
weaknesses, that's easyer to tell. i've not managed erros of the the API calls, the css looks pretty basic and that could definitivly been improved. Also, i had no time to go really indeep in the drag n drop logic as the logic part comes directly from the DnD examples.

> You can use [line comments](https://help.github.com/en/articles/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) to point
to certain parts in your code.
